src/app/clusters/access-control-server/access-control-server.cpp

```diff
@@ -15,4 +15,423 @@
+                            }
+                            break;
+                        case to_underlying(TargetFields::kDeviceType):
+                            if (aReader.GetType() != TLV::kTLVType_Null)
+                            {
+                                ReturnErrorOnFailure(DataModel::Decode(aReader, target.deviceType));
+                                target.flags |= target.kDeviceType;
+                            }
+                            break;
+                        default:
+                            break;
+                        }
+                    }
+                    VerifyOrReturnError(err == CHIP_END_OF_TLV, err);
+                    ReturnErrorOnFailure(aReader.ExitContainer(targetContainer));
+                    ReturnErrorOnFailure(entry.AddTarget(nullptr, target));
+                }
+                VerifyOrReturnError(err == CHIP_END_OF_TLV, err);
+                ReturnErrorOnFailure(aReader.ExitContainer(targetsContainer));
+                break;
+            }
+            default:
+                break;
+            }
+        }
+        VerifyOrReturnError(err == CHIP_END_OF_TLV, err);
+        ReturnErrorOnFailure(aReader.ExitContainer(accessControlEntryContainer));
+        return CHIP_NO_ERROR;
+    }
+
+    AccessControl::Entry entry;
+};
+
+class AccessControlAttribute : public chip::app::AttributeAccessInterface
+{
+public:
+    AccessControlAttribute() : AttributeAccessInterface(Optional<EndpointId>(0), Clusters::AccessControl::Id) {}
+
+    CHIP_ERROR Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder) override;
+    CHIP_ERROR Write(const ConcreteDataAttributePath & aPath, AttributeValueDecoder & aDecoder) override;
+
+private:
+    CHIP_ERROR ReadAcl(AttributeValueEncoder & aEncoder);
+    CHIP_ERROR ReadExtension(AttributeValueEncoder & aEncoder);
+    CHIP_ERROR WriteAcl(AttributeValueDecoder & aDecoder);
+    CHIP_ERROR WriteExtension(AttributeValueDecoder & aDecoder);
+};
+
+CHIP_ERROR AccessControlAttribute::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    switch (aPath.mAttributeId)
+    {
+    case Clusters::AccessControl::Attributes::Acl::Id:
+        return ReadAcl(aEncoder);
+    case Clusters::AccessControl::Attributes::Extension::Id:
+        return ReadExtension(aEncoder);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR AccessControlAttribute::ReadAcl(AttributeValueEncoder & aEncoder)
+{
+    AccessControlEntryCodec codec;
+    AccessControl::EntryIterator iterator;
+
+    GetAccessControl().Entries(iterator);
+
+    CHIP_ERROR err = aEncoder.EncodeList([&](const auto & encoder) -> CHIP_ERROR {
+        while (iterator.Next(codec.entry) == CHIP_NO_ERROR)
+        {
+            encoder.Encode(codec);
+        }
+        return CHIP_NO_ERROR;
+    });
+
+    return err;
+}
+
+CHIP_ERROR AccessControlAttribute::ReadExtension(AttributeValueEncoder & aEncoder)
+{
+    return CHIP_NO_ERROR;
```

Comment:
How about we explicitly encode an empty list and not bloat our RAM use?

Reply:
I can spin a PR for that, or change the XML length to 1 and reduce the octstr size, until we have a real implementation. Which is preferable?

Reply:
Please see PR #12690